### DISCUSSION TITLE
Using the generated client d.ts types

### DISCUSSION THREAD

**Discussion by Twiggeh** [prisma/prisma#7255] [Q&A] [answered]
I tried importing the types out of the d.ts files that the Prisma client generates, but seems to be no way to do that. 

So far I have been inferring the types from functions like `createMany` but that leads to issues like an array being in a Union with a non Array type which needs even further filtering. 

Is there a way to import the types directly ? `import type {stuff} from "./prisma/generated/client/index"` doesn't work. 

Example : 

I have an expensive data transformation, that is clustered. This is the message that the forks receive to work on. I couldn't find a way to import the input / return type of the findMany. I also cannot export the typeof from within the function, so I pulled out everything that returns the types that I want into a seperate function and used `RetrunType` on it

```ts
type workerMessage = readonly [
	Await<
		ReturnType<
			Await<ReturnType<typeof connectToDbs>>['local']['articles']['findMany']
		>
	>,
	Await<
		ReturnType<
			Await<ReturnType<typeof connectToDbs>>['local']['translations']['findMany']
		>
	>,
	Await<
		ReturnType<Await<ReturnType<typeof connectToDbs>>['local']['languages']['findMany']>
	>,
	number
];

// The function that provides the type exports
const connectToDbs = async () => {
	const { PrismaClient: RemotePrismaClient} = await import(
		'../prisma/generated/remote/index.js'
	);
	const { PrismaClient: LocalPrismaClient} = await import(
		'../prisma/generated/local/index.js'
	);

	return {
		remote: new RemotePrismaClient(),
		local: new LocalPrismaClient(),
	};
}
```

This is how I extracted the `CreateManyArticlesInput `, despite it being generated in the `d.ts` file from which I couldn't find a way to import ...
```ts
type CreateManyArticlesInput = NonNullable<
	ArgumentType<
		Await<ReturnType<typeof connectToDbs>>['remote']['articles']['createMany']
	>['0']
>['data'];
```

**Comment by ryands17** [comment]
@Twiggeh 👋 

Could you provide an example of what types would you like to import?

**Reply by Twiggeh** [reply]
Added an example of what I am actually doing :)

**Reply by ryands17** [reply]
This should work:

```ts
import { Prisma, PrismaClient } from '@prisma/client'

type T = Prisma.UserCreateManyInput
```

Given the schema:

```prisma
model User {
  id String @id @default(cuid())
}
```

**Reply by Twiggeh** [reply]
![breh1](https://user-images.githubusercontent.com/23149166/119522656-731ff280-bd7c-11eb-9851-e4cba54890e1.png)

(black bars because of NDA)
Doesn't seem to work. I have multiple servers I am connecting to hence why I am importing the client from a path.

**Reply by ryands17** [reply]
Why are you importing it from `index.js`? Why not just `remote/client'? Also as you're using TypeScript, this should be the recommended way of importing. The `.js` file will not contain the types.

**Reply by Twiggeh** [reply]
It doesn't matter whether I import from `index.js` or not, the error stays the same :

```ts
import type { Prisma } from '../prisma/generated/remote_[redacted]_client/';

const stillFails: Prisma.languagesFindManyArgs = {};
```
```
Namespace '"/mnt/Storage/projects/Webdev/Work/[redacted]/application/data_import/server/prisma/generated/remote_[redacted]_client/index".Prisma' has no exported member 'languagesFindManyArgs'.ts(2694)
```

**Reply by ryands17** [reply]
Have you run `prisma generate` for that specific client? It should show files like these with their type definition:

![image](https://user-images.githubusercontent.com/19697099/119635054-f0a53a80-be30-11eb-929b-991947e9bfcb.png)

Here's an example for the same that you can have a look at:
[test-prisma.zip](https://github.com/prisma/prisma/files/6545389/test-prisma.zip)

**Reply by Twiggeh** [reply]
Yeah, I have, but not under .prisma/client since I have multiple databases that I am connecting to 
![image](https://user-images.githubusercontent.com/23149166/119641087-9351bf00-be19-11eb-93e4-1257dff9c573.png)


I do have access to the types through referencing the `Return<>` and `ArgumentType<>` typescript helper types => I wouldn't have these if the types didn't exist at all.

**Comment by hantsy** [comment]
I have the similar problem when working the generated client. 

I followed the getting started guide, and in my nestjs example, I added a user model same as the  **nestjs** doc , and tried to add a `UserService`(copy the codes  in the page)

```typescript
import {
  UserUpdateInput,
  User,
  UserCreateInput,
  UserWhereUniqueInput,
  UserOrderByInput,
  UserWhereInput,
} from '@prisma/client';
```

The above imports, only User is no errors, others reports no exported members. 

I have checked the `node_modules/.prisma/client/index.d.ts`, the `UserCreateInput` is existed.

**Reply by ryands17** [reply]
It's under the `Prisma` namespace. So this will work:

```ts
import { Prisma } from '@prisma/client'

type T = Prisma.UserUpdateInput
```

**Reply by hantsy** [reply]
Jus check and realize there is a namespace. I think the **nestjs** page should update to it.

**Reply by hantsy** [reply]
@ryands17 Thanks.

**Reply by janpio** [reply]
Which Nestjs page do you refer to exactly? We totally want to update that of course as this shows broken code it seems.

**Reply by hantsy** [reply]
@janpio  https://www.prisma.io/nestjs

**Reply by hantsy** [reply]
![Image 1](https://user-images.githubusercontent.com/854520/119758849-cfe4ef80-bed9-11eb-8a65-78ceeb9212b5.png)

**Reply by Jolg42** [reply, marked as answer]
Indeed! We'll fix that, you can find the updated code at https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nestjs/src for now.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
